lib/Backend/LowerMDShared.cpp

```diff
@@ -8683,24 +8687,21 @@ void LowererMD::GenerateFastInlineBuiltInCall(IR::Instr* instr, IR::JnHelperMeth
             //     if (shouldCheckNegZero) {
             //         CMP roundedFloat, 0
             //         JA $setZero
-            //         JE $negZeroTest
-            //         J $bailout
+            //         JP $skipRoundSd (NaN)
+            //       $negZeroTest [Helper]:
+            //         JNE $bailoutLabel
+            //         isNegZero(src)
+            //         JE $bailoutLabel
+            //         J $skipRoundSd
             //     } else {
             //         CMP roundedFloat, 0
-            //         JL $bailout
-            //         J $setZero
-            //     }
+            //         JNE $setZero
+            //         J $skipRoundSd (0 or NaN)
+            //      }
             // $ltNegHalf:
             //     CMP roundedFloat, NegTwoToFraction
             //     JA $addHalfToRoundSrc
             //     J $skipRoundSd
-            //     if (shouldCheckNegZero) {
-            // $negZeroTest:
-            //         if isNegZero(roundedFloat):
-            //             J $bailout
```

Comment:
The problem is when we specialize result of Math.round to an int, we need to bailout for negative 0, because the conversion to int32 loses that information. With that in mind, I think we can avoid checks for >2^52 as well in case we are specializing the dst to int32 (because we will bailout on overflow).